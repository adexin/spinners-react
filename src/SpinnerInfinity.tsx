import React, { CSSProperties } from 'react';
import { secondaryColorDefaultProps, SecondaryColorSpinnerProps } from './helpers';
import { SpinnersProps, withSharedProps } from './withSharedProps';

import './SpinnerInfinity.css';

export type SpinnerInfinityProps = SpinnersProps & SecondaryColorSpinnerProps;

const Component = ({
  secondaryColor = secondaryColorDefaultProps.secondaryColor,
  speed = secondaryColorDefaultProps.speed,
  still = secondaryColorDefaultProps.still,
  thickness = secondaryColorDefaultProps.thickness,
  ...svgProps
}: SecondaryColorSpinnerProps) => {
  const strokeWidth = 7 * (thickness / 100);
  const dashStyle: CSSProperties = !still
    ? { animation: `spinners-react-infinity ${140 / speed}s linear infinite` }
    : {};

  return (
    <svg fill="none" viewBox="0 0 131 55" {...svgProps}>
      <defs>
        <path
          d="M46.57 45.5138C36.346 55.4954 19.8919 55.4954 9.66794 45.5138C-0.55598 35.5321 -0.55598 19.4678 9.66794 9.48624C19.8919 -0.495412 36.346 -0.495412 46.57 9.48624L84.4303 45.5138C94.6543 55.4954 111.108 55.4954 121.332 45.5138C131.556 35.5321 131.556 19.4678 121.332 9.48624C111.108 -0.495412 94.6543 -0.495412 84.4303 9.48624L46.57 45.5138Z"
          id="spinners-react-infinity-path"
        />
      </defs>
      <use stroke={secondaryColor} strokeWidth={strokeWidth} xlinkHref="#spinners-react-infinity-path" />
      <use
        fill="none"
        stroke="currentColor"
        strokeDasharray="1, 347"
        strokeDashoffset="75"
        strokeLinecap="round"
        strokeWidth={strokeWidth}
        style={dashStyle}
        xlinkHref="#spinners-react-infinity-path"
      />
    </svg>
  );
};

export const SpinnerInfinity = withSharedProps(Component);
