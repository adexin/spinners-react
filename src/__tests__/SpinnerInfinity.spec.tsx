import { render } from '@testing-library/react';
import React from 'react';

import { SpinnerInfinity } from '../SpinnerInfinity';

describe('SpinnerInfinity', () => {
  const color = 'red';

  it('matches the snapshot', () => {
    const { container } = render(
      <SpinnerInfinity color="#fff" size={100} speed={10} thickness={50} />,
    );

    expect(container).toMatchSnapshot();
  });

  it('matches the snapshot with default props', () => {
    const { container } = render(<SpinnerInfinity />);

    expect(container).toMatchSnapshot();
  });

  it('renders null if disabled', () => {
    const { container } = render(<SpinnerInfinity enabled={false} />);

    expect(container.firstChild).toBe(null);
  });

  it('renders still if specified', () => {
    const { container } = render(<SpinnerInfinity still />);
    const uses = container.querySelectorAll('use');

    expect(uses[1].style.animation).toBe('');
  });

  it('passes props to styles', () => {
    const size = '20%';
    const thickness = 40;
    const speed = 50;
    const { container } = render(
      <SpinnerInfinity
        color={color}
        secondaryColor="blue"
        size={size}
        speed={speed}
        thickness={thickness}
      />,
    );
    const uses = container.querySelectorAll('use');
    const strokeWidth = `${7 * (thickness / 100)}`;

    expect(container.firstChild).toHaveStyle({ width: size });
    expect(uses[0]).toHaveAttribute('stroke-width', strokeWidth);
    expect(uses[1]).toHaveAttribute('stroke-width', strokeWidth);
    expect(uses[1].style.animation.includes(`${140 / speed}`)).toBe(true);
  });

  it('passes svg props overriding styles', () => {
    const className = 'test-class';
    const { container } = render(
      <SpinnerInfinity className={className} color="green" style={{ color }} />,
    );

    expect(container.firstChild).toHaveClass(className);
    expect(container.firstChild).toHaveStyle({ color });
  });
});
