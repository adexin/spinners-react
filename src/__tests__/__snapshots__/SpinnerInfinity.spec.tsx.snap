// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpinnerInfinity matches the snapshot 1`] = `
<div>
  <svg
    fill="none"
    style="color: rgb(255, 255, 255); overflow: visible; width: 100px;"
    viewBox="0 0 131 55"
  >
    <defs>
      <path
        d="M46.57 45.5138C36.346 55.4954 19.8919 55.4954 9.66794 45.5138C-0.55598 35.5321 -0.55598 19.4678 9.66794 9.48624C19.8919 -0.495412 36.346 -0.495412 46.57 9.48624L84.4303 45.5138C94.6543 55.4954 111.108 55.4954 121.332 45.5138C131.556 35.5321 131.556 19.4678 121.332 9.48624C111.108 -0.495412 94.6543 -0.495412 84.4303 9.48624L46.57 45.5138Z"
        id="spinners-react-infinity-path"
      />
    </defs>
    <use
      stroke="rgba(0,0,0,0.44)"
      stroke-width="3.5"
      xlink:href="#spinners-react-infinity-path"
    />
    <use
      fill="none"
      stroke="currentColor"
      stroke-dasharray="1, 347"
      stroke-dashoffset="75"
      stroke-linecap="round"
      stroke-width="3.5"
      style="animation: spinners-react-infinity 14s linear infinite;"
      xlink:href="#spinners-react-infinity-path"
    />
  </svg>
</div>
`;

exports[`SpinnerInfinity matches the snapshot with default props 1`] = `
<div>
  <svg
    fill="none"
    style="color: rgb(56, 173, 72); overflow: visible; width: 50px;"
    viewBox="0 0 131 55"
  >
    <defs>
      <path
        d="M46.57 45.5138C36.346 55.4954 19.8919 55.4954 9.66794 45.5138C-0.55598 35.5321 -0.55598 19.4678 9.66794 9.48624C19.8919 -0.495412 36.346 -0.495412 46.57 9.48624L84.4303 45.5138C94.6543 55.4954 111.108 55.4954 121.332 45.5138C131.556 35.5321 131.556 19.4678 121.332 9.48624C111.108 -0.495412 94.6543 -0.495412 84.4303 9.48624L46.57 45.5138Z"
        id="spinners-react-infinity-path"
      />
    </defs>
    <use
      stroke="rgba(0,0,0,0.44)"
      stroke-width="7"
      xlink:href="#spinners-react-infinity-path"
    />
    <use
      fill="none"
      stroke="currentColor"
      stroke-dasharray="1, 347"
      stroke-dashoffset="75"
      stroke-linecap="round"
      stroke-width="7"
      style="animation: spinners-react-infinity 1.4s linear infinite;"
      xlink:href="#spinners-react-infinity-path"
    />
  </svg>
</div>
`;
